sensing/pointcloud_preprocessor/include/pointcloud_preprocessor/distortion_corrector/distortion_corrector.hpp

```diff
@@ -33,54 +35,133 @@
+  virtual void setPointCloudTransform(
+    const std::string & base_link_frame, const std::string & lidar_frame) = 0;
+  virtual void initialize() = 0;
+  virtual void undistortPointCloud(bool use_imu, sensor_msgs::msg::PointCloud2 & pointcloud) = 0;
+};
 
-private:
-  void onPointCloud(PointCloud2::UniquePtr points_msg);
-  void onTwistWithCovarianceStamped(
-    const geometry_msgs::msg::TwistWithCovarianceStamped::ConstSharedPtr twist_msg);
-  void onImu(const sensor_msgs::msg::Imu::ConstSharedPtr imu_msg);
-  bool getTransform(
-    const std::string & target_frame, const std::string & source_frame,
-    tf2::Transform * tf2_transform_ptr);
+template <class Derived>
+class DistortionCorrector : public DistortionCorrectorBase
+{
+public:
+  bool pointcloud_transform_needed_{false};
+  bool pointcloud_transform_exists_{false};
+  bool imu_transform_exists_{false};
+  rclcpp::Node * node_;
+  tf2_ros::Buffer tf_buffer_;
+  tf2_ros::TransformListener tf_listener_;
 
-  bool undistortPointCloud(const tf2::Transform & tf2_base_link_to_sensor, PointCloud2 & points);
+  std::deque<geometry_msgs::msg::TwistStamped> twist_queue_;
+  std::deque<geometry_msgs::msg::Vector3Stamped> angular_velocity_queue_;
+
+  explicit DistortionCorrector(rclcpp::Node * node)
+  : node_(node), tf_buffer_(node_->get_clock()), tf_listener_(tf_buffer_)
+  {
+  }
+  void processTwistMessage(
+    const geometry_msgs::msg::TwistWithCovarianceStamped::ConstSharedPtr twist_msg) override;
+
+  void processIMUMessage(
+    const std::string & base_link_frame,
+    const sensor_msgs::msg::Imu::ConstSharedPtr imu_msg) override;
+  void getIMUTransformation(
+    const std::string & base_link_frame, const std::string & imu_frame,
+    geometry_msgs::msg::TransformStamped::SharedPtr geometry_imu_to_base_link_ptr);
+  void storeIMUToQueue(
+    const sensor_msgs::msg::Imu::ConstSharedPtr imu_msg,
+    geometry_msgs::msg::TransformStamped::SharedPtr geometry_imu_to_base_link_ptr);
+
+  bool isInputValid(sensor_msgs::msg::PointCloud2 & pointcloud);
+  void getIteratorOfTwistAndIMU(
+    bool use_imu, double first_point_time_stamp_sec,
+    std::deque<geometry_msgs::msg::TwistStamped>::iterator & it_twist,
+    std::deque<geometry_msgs::msg::Vector3Stamped>::iterator & it_imu);
+  void undistortPointCloud(bool use_imu, sensor_msgs::msg::PointCloud2 & pointcloud) override;
+  void warnIfTimestampsTooLate(bool is_twist_time_stamp_too_late, bool is_imu_time_stamp_too_late);
+  void undistortPoint(
+    sensor_msgs::PointCloud2Iterator<float> & it_x, sensor_msgs::PointCloud2Iterator<float> & it_y,
+    sensor_msgs::PointCloud2Iterator<float> & it_z,
+    std::deque<geometry_msgs::msg::TwistStamped>::iterator & it_twist,
+    std::deque<geometry_msgs::msg::Vector3Stamped>::iterator & it_imu, float & time_offset,
+    bool & is_twist_valid, bool & is_imu_valid)
+  {
+    static_cast<Derived *>(this)->undistortPointImplementation(
```

Comment:
this one being inlined is critical for performance.
I assume you did, but did you verify that is indeed inlined? I know how it works now, but would it be a problem adding the keyword just to make it explicit?

Reply:
Thanks for the suggestion!

I think CRTP and inline aim for different speeds up.
CRTP:  reduce runtime overhead by avoiding virtual function calls.
inline: insert the function's code directly at each call site.

From my understanding, they could work together. Therefore I add the inline in https://github.com/autowarefoundation/autoware.universe/pull/7137/commits/39f4f629a3ac624bc8e89e3ee72b0f085946f105

Reply:
I asked you in the office to confirm if it was actually being inlined. Did you check? Did you confirm speedups in each case just to know if there is a noticeable difference?

I asked you way before to have a proper processing time evaluation. This is a little more important now that before, since 
 the PR before was the same code base with a new flag, but now there is a complete refactor. The number of frames you checked seemed quite low. While doing that, please fix the units in your measurements

Reply:
> I asked you in the office to confirm if it was being inlined. Did you check?
Do you mean I should check the assembly code to make sure? or do you suggest another way to check it?

I found that I input the inline in the wrong place last time so I fixed it in https://github.com/autowarefoundation/autoware.universe/pull/7137/commits/76603ebb366e817f51b33d32141a831d4a926f6d. And also make the speed difference comparison below https://github.com/autowarefoundation/autoware.universe/pull/7137#issuecomment-2172542605

Reply:
@vividf 
I think everyone has their own way to check these things. Personally, I just use `objdump -TC the_library_name.so | grep ...`. If you have other ways, those should be fine as well.

Judging from your results, it looks they should be the same which is quite welcome. Just for future reference, please add a description on how you measure those times and on what you measured them (to be able to pinpoint any procedural or data issues)

Reply:
For the first one, I try with
```
objdump -d -TC build/pointcloud_preprocessor/CMakeFiles/distortion_corrector_node.dir/rclcpp_components/node_main_distortion_corrector_node.cpp.o
```
but couldn't check it. I will go to the office and ask in these days.

For the second one (future reference), I added in the comment below.

Reply:
If you see the CMake,

```
ament_auto_add_library(pointcloud_preprocessor_filter SHARED
  ...
  src/distortion_corrector/distortion_corrector.cpp
  ...
)
```

The symbols are in that target, not on the node ones (haven't tested myself though).

About the timing, that is not enough at times. You should specify the start and end points of the timer, since for example, putting them after the publish makes it vulnerable to code that you can not control

Reply:
> Start from [line 169](https://github.com/vividf/autoware.universe/blob/76603ebb366e817f51b33d32141a831d4a926f6d/sensing/pointcloud_preprocessor/src/distortion_corrector/distortion_corrector.cpp#L169) to [229](https://github.com/vividf/autoware.universe/blob/76603ebb366e817f51b33d32141a831d4a926f6d/sensing/pointcloud_preprocessor/src/distortion_corrector/distortion_corrector.cpp#L229) by using std::chrono::high_resolution_clock::now()

The start and the end points is there.

Reply:
@vividf 
The symbols question is still missing

Reply:
@knzo25 
> The symbols are in that target, not on the node ones

Thanks, I checked it with 
`
colcon build --symlink-install --cmake-args -DCMAKE_BUILD_TYPE=RelWi\ithDebInfo --packages-select pointcloud_preprocessor

objdump -d -TC build/pointcloud_preprocessor/CMakeFiles/pointcloud_preprocessor_filter.dir/src/distortion_corrector/distortion_corrector.cpp.o`

There is no `undistortPoint` and `undistortPointImplementation` in the block of `undistortPointCloud`, which mean it did inlined.